### PR TITLE
feat: support gateway/tcproute

### PR DESCRIPTION
**Pull request by AlinsRan**
### Type of change:

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### What this PR does / why we need it:

* #644 
### Pre-submission checklist:

* [ ] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

**Comment by tao12345666333** [comment]
please resolve conflics. thanks!

**Comment by tao12345666333** [comment]
please resolve conflics again. thanks

**Review by tao12345666333 — COMMENTED** [review]
Can you write about your plans for TCPRoute support in the description of this PR? 
I see that the Add event is mostly handled in this PR, but the Update and Delete events etc. are not handled.

I added the `round 1` flag to the title

**Comment by AlinsRan** [comment]
> Can you write about your plans for TCPRoute support in the description of this PR? I see that the Add event is mostly handled in this PR, but the Update and Delete events etc. are not handled.
> 
> I added the `round 1` flag to the title

Preliminary support update and delete has been given.

**Comment by tao12345666333** [comment]
@AlinsRan could you please merge latest code?

File: pkg/providers/gateway/translation/gateway_tcproute_test.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+	assert.Nil(t, err)
+
+	assert.Equal(t, 1, len(tctx.StreamRoutes))
+	assert.Equal(t, 1, len(tctx.Upstreams))
+
+	r := tctx.StreamRoutes[0]
+	u := tctx.Upstreams[0]
+
+	// Metadata
+	// FIXME
```
Comment (lingsamuel) [inline pkg/providers/gateway/translation/gateway_tcproute_test.go:191]: Fix what?

File: pkg/providers/gateway/translation/gateway_tcproute_test.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	go svcInformer.Run(stopCh)
+	go epInformer.Run(stopCh)
+	cache.WaitForCacheSync(stopCh, svcInformer.HasSynced)
+	cache.WaitForCacheSync(stopCh, epInformer.HasSynced)
+
+	return tr, processCh
+}
+
+func TestTranslateGatewayTCPRoute(t *testing.T) {
+	refKind := func(str gatewayv1alpha2.Kind) *gatewayv1alpha2.Kind {
```
Comment (lingsamuel) [inline pkg/providers/gateway/translation/gateway_tcproute_test.go:141]: good news: we have generic now

Comment (AlinsRan) [reply pkg/providers/gateway/translation/gateway_tcproute_test.go:141]: So? What shall I do?

File: pkg/providers/gateway/translation/gateway_tcproute_test.go (hunk 1 above)
Comment (AlinsRan) [reply pkg/providers/gateway/translation/gateway_tcproute_test.go:191]: Er, it was copied from httproute. I was negligent.

File: pkg/providers/gateway/translation/gateway_tcproute_test.go (hunk 2 above)
Comment (lingsamuel) [reply pkg/providers/gateway/translation/gateway_tcproute_test.go:141]: https://github.com/apache/apisix-ingress-controller/blob/main/pkg/providers/apisix/translation/apisix_route_test.go#L500-L500

Comment (AlinsRan) [reply pkg/providers/gateway/translation/gateway_tcproute_test.go:141]: Oh. Great!

File: pkg/providers/gateway/translation/gateway_tcproute.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+	for i, rule := range tcpRoute.Spec.Rules {
+		for j, backend := range rule.BackendRefs {
+			if backend.Namespace != nil {
+				ns = string(*backend.Namespace)
+			} else {
+				ns = tcpRoute.Namespace
+			}
+			sr := apisixv1.NewDefaultStreamRoute()
+			name := apisixv1.ComposeStreamRouteName(tcpRoute.Namespace, tcpRoute.Name, fmt.Sprintf("%d-%d", i, j))
```
[the remark this answers is not in the document]
Comment (AlinsRan) [reply pkg/providers/gateway/translation/gateway_tcproute.go]: I think it is more meaningful than `j`.

**Review by lingsamuel — APPROVED** [review]

File: test/e2e/suite-gateway/gateway_tcproute.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+		_ = s.NewAPISIXClientWithTCPProxy().
+			GET("/get").
+			Expect().
+			Status(http.StatusOK)
+
+		assert.Nil(ginkgo.GinkgoT(), s.DeleteResourceFromString(tcpRoute), "deleting TCPRoute")
+		assert.Nil(ginkgo.GinkgoT(), s.EnsureNumApisixStreamRoutesCreated(0), "The number of stream_routes should be 0")
+		assert.Nil(ginkgo.GinkgoT(), s.EnsureNumApisixUpstreamsCreated(0), "The number of upstreams should be 0")
+	})
+})
```
Comment (tao12345666333) [inline test/e2e/suite-gateway/gateway_tcproute.go:123]: This case is wrong

**Review by tao12345666333 — CHANGES_REQUESTED** [review]

**Review by tao12345666333 — APPROVED** [review]
LGTM

**Pull request merged**